src/Process/Runner/MutationTestingRunner.php

```diff
@@ -58,45 +60,82 @@ final class MutationTestingRunner
         $this->eventDispatcher = $eventDispatcher;
+        $this->runConcurrently = $runConcurrently;
     }
 
     /**
      * @param Mutation[] $mutations
      */
-    public function run(array $mutations, int $threadCount, string $testFrameworkExtraOptions): void
+    public function run(iterable $mutations, int $threadCount, string $testFrameworkExtraOptions): void
     {
-        $this->eventDispatcher->dispatch(new MutantsCreationWasStarted(count($mutations)));
+        $numberOfMutants = $this->bufferAndCountIfNeeded($mutations);
 
-        $processes = array_map(
-            function (Mutation $mutation) use ($testFrameworkExtraOptions): MutantProcess {
-                $mutant = $this->mutantFactory->create($mutation);
+        $this->eventDispatcher->dispatch(new MutantsCreationWasStarted($numberOfMutants));
 
-                $process = $this->processBuilder->createProcessForMutant($mutant, $testFrameworkExtraOptions);
+        $processes = take($mutations);
+        $processes->map(function (Mutation $mutation) use ($testFrameworkExtraOptions): MutantProcess {
+            $mutant = $this->mutantFactory->create($mutation);
 
-                $this->eventDispatcher->dispatch(new MutantWasCreated());
+            $process = $this->processBuilder->createProcessForMutant($mutant, $testFrameworkExtraOptions);
 
-                return $process;
-            },
-            $mutations
-        );
+            $this->eventDispatcher->dispatch(new MutantWasCreated());
+
+            return $process;
+        });
 
         $this->eventDispatcher->dispatch(new MutantsCreationWasFinished());
+        $this->eventDispatcher->dispatch(new MutationTestingWasStarted($numberOfMutants));
+
+        // We filter these here because it is beyond responsibilities of a process manager to care about types of processes
+        $processes = take($processes)->filter(function (MutantProcess $mutantProcess) {
+            if ($mutantProcess->getMutant()->isCoveredByTest()) {
+                return true;
+            }
+
+            $this->eventDispatcher->dispatch(new MutantProcessWasFinished(
+                MutantExecutionResult::createFromProcess($mutantProcess)
+            ));
 
-        $this->eventDispatcher->dispatch(new MutationTestingWasStarted(count($processes)));
+            return false;
+        });
 
         $this->parallelProcessManager->run($processes, $threadCount);
 
         $this->eventDispatcher->dispatch(new MutationTestingWasFinished());
     }
+
+    /**
+     * @param iterable|mixed[] $subjects
+     */
+    private function bufferAndCountIfNeeded(iterable &$subjects): int
+    {
+        if ($this->runConcurrently) {
+            return 0;
+        }
+
+        $buffer = [];
+
+        // iterator_to_array wants \Traversable, we can have anything else
+        foreach ($subjects as $subject) {
+            $buffer[] = $subject;
+            // TODO in PHP 7.4 use [...$subjects];
+        }
+
+        $subjects = $buffer;
```

Comment:
why is that step necessary? is it because the iterator might not be rewindable?

Reply:
It might not be an iterator. We're given an iterable, which can be any of the things.

Reply:
I mean why is it necessary to convert `iterable $mutations` into `array $mutations` here

Reply:
We need to count them to display the number in the progress bar.

Reply:
but is the assignment necessary (which is what requires a pass by reference)? you are already returning the `count()`

Reply:
Returning to previous question: yes, we have several non-rewindable generators upstream.